### PR TITLE
[WFCORE-2777] credential-store operations instead of alias resources

### PR DESCRIPTION
**Pull request by thaarok**
https://issues.jboss.org/browse/WFCORE-2777 / https://issues.jboss.org/browse/JBEAP-10799

need coordiantion with https://github.com/wildfly-security-incubator/wildfly/pull/202

### CODE REVIEW

**Review by kabir — APPROVED** [review]
Approving paperwork

File: elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 1]
```diff
@@ -154,10 +164,53 @@
...
+    static final SimpleAttributeDefinition SECRET_VALUE = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.SECRET_VALUE, ModelType.STRING, true)
+            .setStorageRuntime()
+            .setMinSize(0)
+            .build();
+
     // Operations
     static final SimpleOperationDefinition RELOAD = new SimpleOperationDefinitionBuilder(ElytronDescriptionConstants.RELOAD, RESOURCE_RESOLVER)
             .build();
 
+    static final SimpleOperationDefinition READ_ALIASES = new SimpleOperationDefinitionBuilder(ElytronDescriptionConstants.READ_ALIASES, RESOURCE_RESOLVER)
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java]: .setRuntimeOnly()

File: elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 2]
```diff
@@ -154,10 +164,53 @@
...
+            .build();
+
+    static final SimpleOperationDefinition ADD_ALIAS = new SimpleOperationDefinitionBuilder(ElytronDescriptionConstants.ADD_ALIAS, RESOURCE_RESOLVER)
+            .setParameters(ALIAS, ENTRY_TYPE, SECRET_VALUE)
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java]: .setRuntimeOnly()

File: elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 3]
```diff
@@ -154,10 +164,53 @@
...
+            .build();
+
+    static final SimpleOperationDefinition REMOVE_ALIAS = new SimpleOperationDefinitionBuilder(ElytronDescriptionConstants.REMOVE_ALIAS, RESOURCE_RESOLVER)
+            .setParameters(ALIAS)
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java]: .setRuntimeOnly()

File: elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 4]
```diff
@@ -154,10 +164,53 @@
...
+            .build();
+
+    static final SimpleOperationDefinition SET_SECRET = new SimpleOperationDefinitionBuilder(ElytronDescriptionConstants.SET_SECRET, RESOURCE_RESOLVER)
+            .setParameters(ALIAS, ENTRY_TYPE, SECRET_VALUE)
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java]: .setRuntimeOnly()

File: elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 5]
```diff
@@ -346,8 +386,76 @@ protected void performRuntime(ModelNode result, OperationContext context, ModelN
                     } catch (StartException e) {
                         throw new OperationFailedException(e);
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java]: This should throw a RuntimeException, not OFE. OFE should only be used for user errors, not server faults.

File: elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 6]
```diff
@@ -346,8 +386,76 @@ protected void performRuntime(ModelNode result, OperationContext context, ModelN
...
                     }
+                    break;
 
+                case ElytronDescriptionConstants.READ_ALIASES:
+                    try {
+                        result.set(credentialStoreService.getValue().getAliases().stream().map(ModelNode::new).collect(Collectors.toList()));
+                    } catch (CredentialStoreException e) {
+                        throw new OperationFailedException(e);
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java]: This should throw a RuntimeException, not OFE. OFE should only be used for user errors, not server faults.

File: elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 7]
```diff
@@ -346,8 +386,76 @@ protected void performRuntime(ModelNode result, OperationContext context, ModelN
...
+                            if (credentialStore.exists(alias, PasswordCredential.class)) {
+                                throw ROOT_LOGGER.credentialAlreadyExists(alias, PasswordCredential.class.getName());
+                            }
+                            storeSecret(credentialStore, alias, secretValue);
+                        } else {
+                            String credentialStoreName = CredentialStoreResourceDefinition.credentialStoreName(operation);
+                            throw ROOT_LOGGER.credentialStoreEntryTypeNotSupported(credentialStoreName, entryType);
+                        }
+                    } catch (CredentialStoreException e) {
+                        throw ROOT_LOGGER.unableToCompleteOperation(e, e.getLocalizedMessage());
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java:420]: This should throw a RuntimeException, not OFE. OFE should only be used for user errors, not server faults.

File: elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 8]
```diff
@@ -346,8 +386,76 @@ protected void performRuntime(ModelNode result, OperationContext context, ModelN
...
+                        try {
+                            credentialStore.flush();
+                        } catch (CredentialStoreException e) {
+                            // the operation fails, return removed entry back to the store to avoid an inconsistency
+                            // between the store on the FS and in the memory
+                            credentialStore.store(alias, retrieved);
+                            throw e;
+                        }
+                    } catch (CredentialStoreException e) {
+                        throw ROOT_LOGGER.unableToCompleteOperation(e, e.getLocalizedMessage());
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java:439]: This should throw a RuntimeException, not OFE. OFE should only be used for user errors, not server faults.

File: elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 9]
```diff
@@ -346,8 +386,76 @@ protected void performRuntime(ModelNode result, OperationContext context, ModelN
...
+                            if ( ! credentialStore.exists(alias, PasswordCredential.class)) {
+                                throw ROOT_LOGGER.credentialDoesNotExists(alias, PasswordCredential.class.getName());
+                            }
+                            storeSecret(credentialStore, alias, secretValue);
+                        } else {
+                            String credentialStoreName = CredentialStoreResourceDefinition.credentialStoreName(operation);
+                            throw ROOT_LOGGER.credentialStoreEntryTypeNotSupported(credentialStoreName, entryType);
+                        }
+                    } catch (CredentialStoreException e) {
+                        throw ROOT_LOGGER.unableToCompleteOperation(e, e.getLocalizedMessage());
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java:460]: This should throw a RuntimeException, not OFE. OFE should only be used for user errors, not server faults.

File: elytron/src/main/java/org/wildfly/extension/elytron/_private/ElytronSubsystemMessages.java [hunk 10]
```diff
@@ -333,6 +333,10 @@
 
     @Message(id = 919, value = "Invalid alias name '%s' because the credential store %s only supports lower case alias names")
     OperationFailedException invalidAliasName(String alias, String credntialStore);
+
+    @Message(id = 920, value = "Credential alias \"%s\" of credential type \"%s\" does not exists in the store")
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/_private/ElytronSubsystemMessages.java]: s/exists/exist

File: elytron/src/main/java/org/wildfly/extension/elytron/_private/ElytronSubsystemMessages.java [hunk 11]
```diff
@@ -333,6 +333,10 @@
...
+    OperationFailedException credentialDoesNotExists(String alias, String credentialType);
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/_private/ElytronSubsystemMessages.java]: s/Exists/Exist

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 12]
```diff
@@ -1219,15 +1219,21 @@ elytron.credential-store.location=File name of credential store storage.
...
-elytron.credential-store.alias=Alias of sensitive information (e.g. password) to be referenced later in configuration.
-elytron.credential-store.alias.add=Add alias to the credential store.
-elytron.credential-store.alias.remove=Remove current alias from the credential store.
-elytron.credential-store.alias.entry-type=Type of an credential (entry) stored in the credential store.
-elytron.credential-store.alias.secret-value=Secret value such as password.
 elytron.credential-store.add=Add another credential store to the configuration.
 elytron.credential-store.remove=Remove credential store from the configuration.
 elytron.credential-store.reload=Reload credential store (reinitialize and load new data if present).
 elytron.credential-store.state=The state of the underlying service that represents this credential store at runtime.
+elytron.credential-store.read-aliases=Read an aliases (entries) from a credential store.
```
Comment (bstansberry) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties]: s/an aliases/aliases/g

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 13]
```diff
@@ -1219,15 +1219,21 @@ elytron.credential-store.location=File name of credential store storage.
...
+elytron.credential-store.add-alias=Add alias to the credential store.
```
Comment (bstansberry) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties]: s/alias/an alias/g

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 14]
```diff
@@ -1219,15 +1219,21 @@ elytron.credential-store.location=File name of credential store storage.
...
+elytron.credential-store.add-alias.alias=The alias of the credential store item to add.
+elytron.credential-store.add-alias.entry-type=Type of an credential (entry) stored in the credential store.
```
Comment (bstansberry) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties]: s/an/a

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 15]
```diff
@@ -1219,15 +1219,21 @@ elytron.credential-store.location=File name of credential store storage.
...
+elytron.credential-store.add-alias.secret-value=Secret value such as password.
```
Comment (bstansberry) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties]: s/password/a password/g

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 16]
```diff
@@ -1219,15 +1219,21 @@ elytron.credential-store.location=File name of credential store storage.
...
+elytron.credential-store.remove-alias=Remove an credential (entry) stored in the credential store.
```
Comment (bstansberry) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties]: s/an/a

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 17]
```diff
@@ -1219,15 +1219,21 @@ elytron.credential-store.location=File name of credential store storage.
...
+elytron.credential-store.remove-alias.alias=The alias of the credential store item to remove.
+elytron.credential-store.set-secret=Set secret value of an credential (entry) stored in the credential store.
```
Comment (bstansberry) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties]: s/secret/the secret/g
s/an/a

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 18]
```diff
@@ -1219,15 +1219,21 @@ elytron.credential-store.location=File name of credential store storage.
...
+elytron.credential-store.set-secret.alias=The alias of the credential store item to set.
+elytron.credential-store.set-secret.entry-type=Type of an credential (entry) stored in the credential store.
```
Comment (bstansberry) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties]: s/an/a

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 19]
```diff
@@ -1219,15 +1219,21 @@ elytron.credential-store.location=File name of credential store storage.
...
+elytron.credential-store.set-secret.secret-value=Secret value such as password.
```
Comment (bstansberry) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties]: s/password/a password/g

**Review by bstansberry — CHANGES_REQUESTED** [review]
I made a ton of comments here but mostly they are the same minor things repeated.

The "throw RuntimeException not OFE" thing is a problem in the subsystem prior to this patch, but this seems like a good time to at least partially correct it.

File: elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java [hunk 20]
```diff
@@ -154,10 +164,53 @@
...
+
+    static final StandardResourceDescriptionResolver RESOURCE_DESCRIPTION_RESOLVER = ElytronExtension.getResourceDescriptionResolver(ElytronDescriptionConstants.CREDENTIAL_STORE, ElytronDescriptionConstants.ALIAS);
+
+    static final SimpleAttributeDefinition SECRET_VALUE = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.SECRET_VALUE, ModelType.STRING, true)
+            .setStorageRuntime()
+            .setMinSize(0)
+            .build();
+
     // Operations
     static final SimpleOperationDefinition RELOAD = new SimpleOperationDefinitionBuilder(ElytronDescriptionConstants.RELOAD, RESOURCE_RESOLVER)
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/CredentialStoreResourceDefinition.java:196]: This one is setRuntimeOnly() as well.

**Comment by thaarok** [comment]
@bstansberry updated by comments

**Comment by honza889** [comment]
ehm... wrong rebase, sorry, will fix soon..

**Comment by bstansberry** [comment]
@honza889 this branch is messed up now. 25 commits.

**Comment by bstansberry** [comment]
@honza889 @darranl the last commit on this branch addresses all my previous comments, so please regard this as having my +1 once the branch is rebased and the other commits removed.

**Comment by darranl** [comment]
Thanks @bstansberry, I am Ok without the rebase, I will be bringing ladybird up to date with master before I merge so git will handle the rest.

**Pull request merged**